src/table/table-adapter.class.ts

```diff
@@ -0,0 +1,184 @@
+	 *
+	 * @param row index of the row
+	 */
+	getRow(row: number): TableRowAdapter { return; }
+
+	/**
+	 * Finds the column index of a given cell
+	 *
+	 * @param cell the cell to search for
+	 */
+	findColumnIndex(cell: TableCellAdapter): number { return; }
+
+	/**
+	 * Finds the row index of a given cell
+	 *
+	 * @param cell the cell to search for
+	 */
+	findRowIndex(cell: TableCellAdapter): number { return; }
+
+	/**
+	 * Finds the row and column index of a given cell
+	 *
+	 * @param cell the cell to search for
+	 * @returns a tuple that follows the `[row, column]` convention
+	 */
+	findIndex(cell: TableCellAdapter): [number, number] { return; }
+}
+
+/**
+ * A concrete implementation of `TableAdapter`
+ *
+ * Provides standard and consistent access to table cells and rows
+ */
+export class TableDomAdapter implements TableAdapter {
+	/**
+	 * The last acccessible column in the table
+	 */
+	public get lastColumn() {
+		return this.tableElement.rows[0].cells.length - 1;
+	}
+
+	/**
+	 * The last accessible row in the table
+	 */
+	public get lastRow() {
+		return this.tableElement.rows.length - 1;
+	}
+
+	/**
+	 * `TableDomAdapter` works on a normal HTML table structure.
+	 * Custom tables that don't follow the standard structure should use a custom implementation of `TableAdapter`.
+	 *
+	 * The standard structure allows us to directly query rows for cells and indexes - though we do have to handle colspans specially.
+	 *
+	 * @param tableElement the root HTML table element.
+	 */
+	constructor(public tableElement: HTMLTableElement) { }
+
+	/**
+	 * Returns a cell from the table taking colspans in to account.
+	 *
+	 * @param row index of the row
+	 * @param column index of the column
+	 */
+	getCell(row: number, column: number): HTMLTableCellElement {
+		const baseRow = this.getRow(row);
+		let baseCells = Array.from(baseRow.cells);
+
+		// rows can have fewer total cells than the actual table
+		// the model pretends all rows behave the same (with colspans > 1 being N cells long)
+		// this maps that view to the HTML view (colspans > 1 are one element, so the array is shorter)
+		let realIndex = 0;
+		// i is only used for iterating the "cells"
+		for (let i = 0; i < column;) {
+			// skip the next N "cells"
+			i += baseCells[realIndex].colSpan;
+			// don't bump realIndex if i now exceeds the column we're shooting for
+			if (i > column) { break; }
+			// finally, increment realIndex (to keep it generally in step with i)
+			realIndex++;
+		}
+		return baseCells[realIndex];
+	}
+
+	/**
+	 * Returns a row from the table
+	 *
+	 * @param row index of the row
+	 */
+	getRow(row: number): HTMLTableRowElement {
+		return this.tableElement.rows[row];
+	}
+
+	/**
+	 * Finds the column index of a given cell
+	 *
+	 * @param cell the cell to search for
+	 */
+	findColumnIndex(cell: HTMLTableCellElement): number {
+		const row = this.getRow(this.findRowIndex(cell));
+		let cellIndex = 0;
+		for (const c of Array.from(row.cells)) {
```

Comment:
Nope, `HTMLTableRowElement` is a `HTMLCollection` not an array, and it isn't `Iterable` ... because reasons 😒

Reply:
What may make more sense is just having the underlying `getRow` return a `TableRowAdapter[]` instead of a `HTMLCollection`

Reply:
Nevermind, I have no idea what I was saying ... That'll have to stay as is - we still return a type compliant with `TableRowAdapter`, but for this case we have the whole element ... I don't think it's a good plan to start returning a more abstract representation at this point (though we could in the future without breaking anyone)

Reply:
That's weird... `for of` _should_ work with `HTMLCollection`s if I'm reading this code correctly http://jsfiddle.net/jfriend00/3ddpz8sp/

Reply:
<img width="1328" alt="image" src="https://user-images.githubusercontent.com/1744185/65621238-5de43680-df91-11e9-87f2-cf700c31c202.png">

It may work in (some) browsers, but TS doesn't support it

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/HTMLCollection

https://developer.mozilla.org/en-US/docs/Web/API/NodeList

https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll

Additionally, `querySelectorAll` returns a `NodeList`, which behaves slightly differently than a `HTMLCollection` ... it also doesn't officially support the `Iterator` protocol. The recommended way to use both in JS is via `Array.from`

Reply:
I see. 👍